sklearn/utils/extmath.py

```diff
@@ -360,6 +356,164 @@ def randomized_svd(M, n_components, n_oversamples=10, n_iter='auto',
+
+    flip_sign : boolean, (default=True)
+        The output of a singular value decomposition is only unique up to a
+        permutation of the signs of the singular vectors. If `flip_sign` is
+        set to `True`, the sign ambiguity is resolved by making the largest
+        loadings for each component in the left singular vectors positive.
+
+    random_state : int, RandomState instance or None, optional (default=None)
+        The seed of the pseudo random number generator to use when shuffling
+        the data.  If int, random_state is the seed used by the random number
+        generator; If RandomState instance, random_state is the random number
+        generator; If None, the random number generator is the RandomState
+        instance used by `np.random`.
+
+    Notes
+    -----
+    This algorithm finds a (usually very good) approximate truncated principal
+    component analysis decomposition using randomized methods to speed up the
+    computations. It is particulary useful on large, sparse matrices since this
+    implementation doesn't require centering the original matrix (which would
+    center and therefore densify potentially large sparse matrices, leading to
+    memory issues). In order to obtain further speed up, `n_iter` can be set
+    <=2 (at the cost of loss of precision).
+
+    References
+    ----------
+    * Algorithm 971: An implementation of a randomized algorithm for principal
+      component analysis
+      Li, Huamin, et al. 2017
+
+    """
+    random_state = check_random_state(random_state)
+
+    if n_iter == "auto":
+        # Checks if the number of iterations is explicitly specified
+        # Adjust n_iter. 7 was found a good compromise for PCA.
+        # See sklearn #5299
+        n_iter = 7 if n_components < .1 * min(A.shape) else 4
+
+    # Deal with "auto" mode
+    if power_iteration_normalizer == "auto":
+        if n_iter <= 2:
+            power_iteration_normalizer = "none"
+        else:
+            power_iteration_normalizer = "LU"
+
+    n_samples, n_features = A.shape
+
+    if sparse.issparse(A):
+        means, _ = mean_variance_axis(A, axis=0)
+    else:
+        means = np.mean(A, axis=0)
+    c = np.atleast_2d(means)
+
+    if n_samples >= n_features:
+        Q = random_state.normal(
+            size=(n_features, n_components + n_oversamples)
+        )
+        if A.dtype.kind == "f":
+            Q = Q.astype(A.dtype, copy=False)
+
+        Q = safe_sparse_dot(A, Q) - safe_sparse_dot(c, Q)
+
+        # Normalized power iterations
+        for _ in range(n_iter):
+            Q = safe_sparse_dot(A.T, Q) - \
+                safe_sparse_dot(c.T, Q.sum(axis=0)[None, :])
+            Q = _normalize_power_iteration(Q, power_iteration_normalizer)
+            Q = safe_sparse_dot(A, Q) - safe_sparse_dot(c, Q)
+            Q = _normalize_power_iteration(Q, power_iteration_normalizer)
+
+        Q, _ = linalg.qr(Q, mode="economic")
+
+        QA = safe_sparse_dot(A.T, Q) - \
```

Comment:
Also please try not to deviate from the `randomized_svd` names because this makes comparison beetween the 2 methods more complicated than it could be.

`QA.T` should be `B`, `R` should be `Uhat` (or `Vhat`), and the matrix `A` should be `M`.